theano/tensor/nnet/tests/test_nnet.py

```diff
@@ -255,14 +293,109 @@ def test_local_softmax_optimization(self):
+
+    def test_local_indexing_softmax_optimization(self):
+        """Test the Logsoftmax substitution with indexing
+
+        Check that Log(Softmax(x)[y]) is substituted with Logsoftmax(x)[y]. Note that
+        only the forward pass is checked (i.e., doesn't check the gradient)
+        """
+        dims = 4
+        # Vector case tested in test_optimize_xent_vector
+        for d in xrange(2, dims + 1):
+            # Case 1: Log(Subtensor(softmax(x)))
+            x = T.TensorType(dtype=config.floatX, broadcastable=(False,) * d)('x')
+            sm = tensor.nnet.softmax(x)[3]
+            logsm = tensor.log(sm)
+            f = theano.function([x], logsm)
+            assert isinstance(f.maker.fgraph.outputs[0].owner.inputs[0].owner.op, theano.tensor.nnet.nnet.LogSoftmax)
+
+            # Case 2: Log(Advanced_Subtensor1(softmax(x)))
+            x = T.TensorType(dtype=config.floatX, broadcastable=(False,) * d)('x')
+            sm = tensor.nnet.softmax(x)[range(1, 3)]
+            logsm = tensor.log(sm)
+            f = theano.function([x], logsm)
+            assert isinstance(f.maker.fgraph.outputs[0].owner.inputs[0].owner.op, theano.tensor.nnet.nnet.LogSoftmax)
+
+        # Case 3: Log(Advanced_Subtensor(softmax(x)))
+        x = T.TensorType(dtype=config.floatX, broadcastable=(False,) * 3)('x')
+        sm = tensor.nnet.softmax(x)[:, range(1, 3), 2]
         logsm = tensor.log(sm)
         f = theano.function([x], logsm)
-        assert isinstance(f.maker.fgraph.outputs[0].owner.op,
-                          theano.tensor.nnet.nnet.LogSoftmax)
-        assert check_stack_trace(
-            f, ops_to_check=theano.tensor.nnet.nnet.LogSoftmax)
+        assert isinstance(f.maker.fgraph.outputs[0].owner.inputs[0].owner.op, theano.tensor.nnet.nnet.LogSoftmax)
+
+    def test_local_indexing_softmax_grad_optimization(self):
+        """Test the Logsoftmax'grad substitution with indexing
+
+        Check that Log(Softmax(x)[y])'s grad' is substituted by the corresponding
+        expression. And check the newly computed grad values with verify_grad.
+        """
+        dims = 4
+        # Vector case tested in test_optimize_xent_vector
+        for d in xrange(2, dims + 1):
+            shape = (5,) * d
+            # Case 0: Log(softmax(x))
+
+            def f(a):
+                return tensor.log(tensor.nnet.softmax(a))
+
+            x = T.TensorType(dtype=config.floatX, broadcastable=(False,) * d)('x')
+            logsm = f(x).sum()
+            g = T.grad(logsm, x)
+            fgraph = gof.FunctionGraph([x], [g])
+            theano.compile.mode.optdb.query(theano.compile.mode.OPT_FAST_RUN).optimize(fgraph)
+            assert softmax_grad not in [n.op for n in fgraph.toposort()]
+            utt.verify_grad(f, [np.random.rand(*shape)])
+
+            # Case 1: Log(Subtensor(softmax(x)))
+
+            def f(a):
+                return tensor.log(tensor.nnet.softmax(a)[3])
```

Comment:
I found an issue in this PR, when we have tensor.log(tensor.nnet.softmax(a)[:, 3]), i.e we get a subtensor on the column instead of the row, we got wrong gradients when we check them with the finite difference methods.
I printed the differences for an example and with the finite differences method, I get:
```
[array([[-0.23060371, -0.15813434, -0.2098594 ,  0.97880861, -0.38021126],
       [-0.3059443 , -0.30146484, -0.21215589,  1.04132029, -0.22175533],
       [-0.1581472 , -0.31918843, -0.27057473,  0.91833941, -0.17042915],
       [-0.27672931, -0.18285466, -0.22751856,  0.92796999, -0.24086756],
       [-0.27860693, -0.27760008, -0.18710686,  1.18787938, -0.44456558]])]
```
And with my optimization, I have:
```
[array([[ 0.        ,  0.        ,  0.        ,  0.01876505,  0.        ],
       [ 0.        ,  0.        ,  0.        , -0.09793607,  0.        ],
       [ 0.        ,  0.        ,  0.        , -0.22800455,  0.        ],
       [ 0.        ,  0.        ,  0.        , -0.33536798,  0.        ],
       [ 0.        ,  0.        ,  0.        ,  0.28916284,  0.        ]])]
```

I think, I will have to check the maths to see if the gradients is effectively defined for all the matrix (even if intuitively it should be defined only on the subtensor and be 0 for the rest).

Reply:
I just check the math. I think there is a problem with the optimization. It should result in non zero gradients for the other columns. The reason is that for example, softmax(a)[0,3] is a function of a[0,:]. I'm going to check the code too.